client/cmd.go

```diff
@@ -80,17 +80,17 @@ func ValidateCmd(cmd *cobra.Command, args []string) error {
 // flags that do not necessarily change with context. These must be checked if
 // the caller explicitly changed the values.
 func ReadPersistentCommandFlags(clientCtx Context, flagSet *pflag.FlagSet) (Context, error) {
-	if flagSet.Changed(cli.OutputFlag) {
+	if clientCtx.OutputFormat == "" || flagSet.Changed(cli.OutputFlag) {
```

Comment:
Why so we need to check both of these? What happens if we just check `clientCtx`?

Reply:
See the PR body as it lays out all 4 possible cases.

If, for example, you _only_ check `clientCtx`, then the following case fails:

4. `client.Context` field pre-populated & flag set -> uses set flag value ❌  (uses pre-populated value instead)

Because if a field in `client.Context` is set (e.g. `Node`) AND the client supplies such a flag, then using `clientCtx` only will not allow the command to read what the client provided. Having both checks allows us to cover all 4 possible cases in a consistent manner.

Reply:
Ah right. I think I get it now.

Reply:
Though I think my proposal of not using flags to set defaults at all would more succinctly have this with just the `Changed` check.

Reply:
Not sure I follow. Have code I can look at?

Reply:
So for most of these cases, the default is actually `false`, `0` or `""`. In those cases there is no need to check if the `clientCtx` value is not default.

i.e. if we check `clientCtx.Offline` and it's false, what difference does it make if we call `flagSet.GetBool(flags.FlagOffline)` just to read the default value of `false` which was the value anyway. This code is equivalent has the exact same behavior as the code in this PR:

```go
if flagSet.Changed(flags.FlagOffline) {
 		offline, _ := flagSet.GetBool(flags.FlagOffline)
 		clientCtx = clientCtx.WithOffline(offline)
 	}
```

There are only a few cases where the flag default value is not empty. `OutputFlag` is one of them. So maybe that's a case where `if clientCtx.OutputFormat == "" || flagSet.Changed(cli.OutputFlag)` is merited. But still I would argue that if the flag isn't set and the default of `"text"` isn't read from `flagSet.GetString(cli.OutputFlag)`, `client.Context` should treat `""` as the default value of `"text"`. It shouldn't rely on `flagSet.GetString` to have the correct default behavior.

Reply:
The idea is the that we have consistent logic for _all_ flags/context fields. You don't have to think if a flag is potentially pre-populated or not or what kind of default value or type it might have. This PR introduces consistent code to cover all 4 possible cases. Though I do agree for some of these flags, it makes no sense to have both checks, but I'd again opt for consistency above all else.

Reply:
I think I get what you're after. Basically flag sets may define defaults that are different from the `client.Context` default, but maybe not. But either way you want to consistently set the flag default as the last fallback if you're reading flags. Is that right?

Reply:
>I think I get what you're after. Basically flag sets may define defaults that are different from the client.Context default, but maybe not.

Yes!

> But either way you want to consistently set the flag default as the last fallback if you're reading flags. Is that right?

Yes!

Reply:
That's fine. It would be great to make sure that intention really clearly documented.

Reply:
I'll document each of the three helper functions.